File: medium/81. Search in Rotated Sorted Array II/81. Search in Rotated Sorted Array II.ts
function search(nums: number[], target: number): boolean {
    let left = 0, right = nums.length - 1

    while (left <= right) {
        const mid = Math.floor((left + right) / 2)

        if (nums[mid] == target)
            return true

        if (nums[left] == nums[mid] && nums[mid] == nums[right]) {
            left += 1
            right -= 1
        } else if (nums[left] <= nums[mid]) {
            if (nums[left] <= target && target < nums[mid]) {
                right = mid - 1
            } else {
                left = mid + 1
            }
        } else {
            if (nums[mid] < target && target <= nums[right]) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
    }

    return false
};

export { }

File: medium/81. Search in Rotated Sorted Array II/81. Search in Rotated Sorted Array II.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        n = len(nums) // 2

        return self.binary_search(nums[:n], target, True) or self.binary_search(
            nums[n:], target, False
        )

    def binary_search(self, nums: List[int], target: int, ifForward: bool) -> bool:
        start = 0 if ifForward else len(nums) - 1
        end = len(nums) if ifForward else -1
        step = 1 if ifForward else -1

        for i in range(start, end, step):
            if nums[i] == target:
                return True

        return False


solution = Solution()
# print(solution.search(nums=[2, 5, 6, 0, 0, 1, 2], target=0))
# print(solution.search(nums=[2, 5, 6, 0, 0, 1, 2], target=3))
# print(solution.search(nums=[1, 0, 1, 1, 1], target=0))
print(
    solution.search(
        nums=[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1], target=2
    )
)
# print(
#     solution.search(
#         nums=[1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], target=2
#     )
# )
